Generate JUnit XML reports in the runner

codeblooded opened the pull request:
This commit adds support for exporting JUnit XML reports from the runner.

jtattermusch commented:
Before merging, please post a resultstore/sponge link here so I can see how the structured test results looks like (and provide suggestions for improvements).

codeblooded commented:
@jtattermusch Here is a resultstore link to go tests that succeeded: https://source.cloud.google.com/results/invocations/75dd8d04-3c13-4720-8fdc-bac026c4a913/targets;collapsed=/github%2Fgrpc/tests. I can generate one with failures in another run. 

Since it is open source, I strictly followed the JUnit schema and did not add any arbitrary fields: https://www.ibm.com/docs/en/z-open-unit-test/2.0.x?topic=SSZHNR_2.0.0/com.ibm.rsar.analysis.codereview.cobol.doc/topics/cac_useresults_junit.html.

jtattermusch commented:
@codeblooded  I looked at the report and it's ok as a start, but it doesn't really provide any insights beyond  "some 32 core scenarios have passed" and "some 8 core scenarios have passed".

- from the name of the scenario I don't know which scenario is which
- ideally I'd like to see some basic log (e.g. the driver's output which is short and shows basic info about what has run and what was the results measured) for both passing and failing scenarios.

![image](https://user-images.githubusercontent.com/9939684/118940570-263db580-b951-11eb-8c89-3bafe41da905.png)


Here's one of the old reports from the legacy framework for inspiration:

Driver's output for failed logs:
https://source.cloud.google.com/results/invocations/2abc6ab4-38ad-4d3a-b195-582e9b2e26f5/targets/github%2Fgrpc%2Freports%2F8core/tests;passed=true;group=benchmarks;test=qps_json_driver.ruby_to_cpp_protobuf_sync_streaming_ping_pong;row=57

nit: I think it also makes more sense to display a separate top-level "target" for 8core and 32core scenario:
e.g. see here: https://source.cloud.google.com/results/invocations/2abc6ab4-38ad-4d3a-b195-582e9b2e26f5/targets;collapsed=  (so that the top level target doesn't just say "github/grpc" - which is not all that informative. IIRC this can be achieved by generating multiple xml files in subdirectories that are named accordingly.

jtattermusch commented:
Btw in the link you sent, the kokoro job has actually failed:
```
panic: Wrote XML report (4329 bytes) to file: sponge_log.xml

goroutine 1 [running]:
log.Panicf(0x124979c, 0x27, 0xc00041ff08, 0x2, 0x2)
	/usr/local/go/src/log/log.go:340 +0xc0
main.main()
	/tmpfs/src/github/test-infra/cmd/runner/main.go:111 +0xe6b
```

jtattermusch commented:
Seeing a report where there are failures seems useful, since this is where the reports will come especially handy (and they should actually be useful when someone wants to look for what the problem was).

codeblooded commented:
@jtattermusch That's my mistake. I was using autocomplete and grabbed `log.Panicf` instead of `log.Printf`, so it showed the correct log and then panicked.

As for the names, you are right. I can modify it to show which scenario it came from. Maybe the `id` can be the name of the test and the `name` could be the scenario name.

Logs for each test case make sense. There is some complication collecting the driver logs since the runner does not actively interact with pods, but I can work on that. This may have some added toil, because there is another bug right now that will change the way I find it. (I'd link but it's internal.) I'll also have to decide whether to use the Kubernetes API or to wrap kubectl. The benefit of kubectl is that I can get all the logs in one go for all pods in the test. Each line would have a prefix saying the timestamp and which pod it came from. That said, I hate wrapping command line tools when there is an API. Sometimes the API is inconvenient though. 

I can create separate targets, but would it only be one `<testsuite>` in the `<testsuites>` then? Is there some other way we'd like to group them? Maybe by language would be a good idea. I'll prioritize the other pieces first and then come back to this.

paulosjca commented:
> As for the names, you are right. I can modify it to show which scenario it came from. Maybe the `id` can be the name of the test and the `name` could be the scenario name.

The runner already has a nameString() function to display the name of the test as it should have been. You can use that to get a readable name for display.

paulosjca commented:
> Logs for each test case make sense. There is some complication collecting the driver logs since the runner does not actively interact with pods, but I can work on that. This may have some added toil, because there is another bug right now that will change the way I find it. (I'd link but it's internal.) I'll also have to decide whether to use the Kubernetes API or to wrap kubectl. The benefit of kubectl is that I can get all the logs in one go for all pods in the test. Each line would have a prefix saying the timestamp and which pod it came from. That said, I hate wrapping command line tools when there is an API. Sometimes the API is inconvenient though.

I think this can be a separate work item, first we generate reports, then we add logs to them.

> I can create separate targets, but would it only be one `<testsuite>` in the `<testsuites>` then? Is there some other way we'd like to group them? Maybe by language would be a good idea. I'll prioritize the other pieces first and then come back to this.

I think one testsuite would be fine, as it is what we had before. The flag for it could be a map aggregator, like the concurrency level (queue:filename).

paulosjca commented:
> Logs for each test case make sense. There is some complication collecting the driver logs since the runner does not actively interact with pods, but I can work on that. This may have some added toil, because there is another bug right now that will change the way I find it. (I'd link but it's internal.) I'll also have to decide whether to use the Kubernetes API or to wrap kubectl. The benefit of kubectl is that I can get all the logs in one go for all pods in the test.

Getting them one by one is ok, since we already make calls to create and poll the tests. it might make sense too if in the future we decide to have an option to delete the test once we collect the logs.

paulosjca commented on inspect/doc.go:
```diff
@@ -0,0 +1,19 @@
+limitations under the License.
+*/
+
+// Package inspect contains utilities for extracting information about a
```
This "inspect" package goes against the design. Any information that is needed to produce reports should be added to the metadata when constructing the test configuration. This is already the case for scenario. Any other informatioin that is needed can be added by the loadtest config generator, either explicitly as a flag or by updating the code to generate it (as already happens with scenario).

We need a design discussion and review before continuing with the implementation of this feature.

paulosjca commented on tools/runner/runner.go:
```diff
@@ -117,14 +120,13 @@ func (r *Runner) runTest(config *grpcv1.LoadTest, reporter *TestCaseReporter, do
 			if retries < r.retries {
 				retries++
-				reporter.Info("Scheduling retry %d/%d to poll test", retries, r.retries)
+				reporter.Info("Failed to poll test %s (scheduling retry %d/%d): %v", name, retries, r.retries, err)
```
It is not necessary to mention the name of the test. I would use comma instead of parentheses.

paulosjca commented on tools/runner/runner.go:
```diff
@@ -117,14 +120,13 @@ func (r *Runner) runTest(config *grpcv1.LoadTest, reporter *TestCaseReporter, do
 				continue
 			}
-			reporter.Error("Aborting test after %d retries to poll test %s: %v", r.retries, name, err)
+			reporter.Error("Error polling the test", "Aborting test after %d retries to poll test %q failed: %v", r.retries, name, err)
```
Generally speaking, it is unnecessary to mention the name of the test. I would not add "failed" either.

paulosjca commented on tools/runner/runner.go:
```diff
@@ -134,7 +136,11 @@ func (r *Runner) runTest(config *grpcv1.LoadTest, reporter *TestCaseReporter, do
 		case loadTest.Status.State.IsTerminated():
-			reporter.Info("%s", status)
+			if status != "Succeeded" {
+				reporter.Error("Test failed", "Test %q failed with reason %q: %v", name, loadTest.Status.Reason, loadTest.Status.Message)
```
I would not mention the name of the test here.

paulosjca commented on cmd/runner/main.go:
```diff
@@ -58,13 +73,33 @@ func main() {
+
+	suitesReporter.SetEndTime(time.Now())
+
+	if len(o) > 0 {
```
There are multiple options here:

1. Should we always generate an xml report, and print it to stdout if no output file is specified?
2. Should we always generate an xml report, and print it to a file only if it is specified?
3. Should we generate an xml report only if an output file is specified?

Right now we are following option 2. It seems wasteful to create a report that is not used.

What we could do instead is create the "suites" report in the main file, and pass it as an argument when creating the suitesReporter. This would trigger the creation of "suite" and "case" reports at the lower levels.

If we do that, we will need to keep a testCaseCount in the suiteReporter, since the suite report itself may be nil and not keep a list of case reports.

We might get maximum flexibility by having an argument to pass in a suiteReport when creating a suiteReporter as well. That way we could explicitly create a report only for testSuites for which a report is requested.

codeblooded replied:
I do not think we should generate an XML report unless it is specified. I believe the driver does not at present. I agree that (2) has unnecessary data structure creation. That is why I'm not sure about the coupling of the logger and the junit report. To me, (3) seems most appropriate.

codeblooded replied:
Choosing (3), see below for implementation discussion and details.

paulosjca commented on cmd/runner/main.go:
```diff
@@ -15,26 +15,41 @@ package main
+
 func main() {
 	var i runner.FileNames
+	var o runner.FileNames
```
If we know we are generating only a single report, then this argument can be a string instead of a list accumulator.

If we want ot specify multiple reports, then we may want to have a map accumulator instead, as we have for concurrency levels. I might rename FileNames to InputFileNames and call the new map accumulator OutputFileNames.

codeblooded replied:
Right, I temporarily used a runner.FileNames since it provides validation. In this case, it requires the string to not be empty. We could create a OutputFileName flag type that validates the location is writable on the flag parse Set call. We could also validate that the files exist for FileNames, making it InputFileNames. Or we can choose to handle the logic later. The flexibility of the flag.Var interface is quite nice.

paulosjca commented on cmd/runner/main.go:
```diff
@@ -15,26 +15,41 @@ package main
 	flag.Var(&c, "c", "concurrency level, in the form [<queue name>:]<concurrency level>")
+	flag.Var(&o, "o", "name of the output file for junit xml report")
 	flag.StringVar(&a, "annotation-key", "pool", "annotation key to parse for queue assignment")
+	flag.StringVar(&outputSuiteName, "output-suite-name", defaultOutputSuiteName, "name field for testsuites in junit xml report")
```
I wonder how useful it is to be able to specify this name. If it can be empty, I would make the default an empty string. If it cannot, I would suggest making this flag required.

If it is useful to specify this name, I would rename the flag to something like "suites-name" (so it is clear it is not the name of a single suite, and it is not something you output, it is something you set). I would change the description to "suites name for junit xml report".

I would also look into the ordering of variable and flag declarations, so they reasonably match each other (i, o, c versiosn i, c, o right now).

codeblooded replied:
Generally it is a best practice to have a id for each XML field. The schema specified here: https://github.com/windyroad/JUnit-Schema/blob/162a883ac631f61ca1d14a3d472e1fbb53c1c05f/JUnit.xsd#L33-L37 does not appear to require one for `<testsuites>` only `<testsuite>`. I do think it will directly correlate with what we output when we modify to produce multiple files like @jtattermusch requested. When this happens, I don't think this flag will be required. I'll leave it for now though.

codeblooded commented:
@jtattermusch Commit  bd2e53e had a rough prototype working, but the team decided that the final design needed to change. Commits were merged into master that required this to be rewritten. We are now trying to figure out what the best design would look like.

codeblooded commented:
Reading over this review, I'm feeling like the JUnit logic is very coupled to the reporter. The ever-present logger is also a reporter. And the reporter is very coupled to the runner. But, you can have a reporter without JUnit. Is this a reporter or a logger with a JUnit report as output?

The reporter is confusing for logs. For example, the `Warning` and `Error` methods must have both a message and a text body for the JUnit reports. However, the message is completely ignored by the stdout log. I see how storing the actual JUnit child objects is useful. But, it breaks thread-safety and they aren't necessary if no JUnit report is being generated.

If we want a JUnit report to be a type of Log (do we?), then it might make sense to decouple the JUnit and standard logging, like-so:

```go
type Logger interface {
  Info(detailsFmt string, v ...interface{})
  Warning(brief, detailsFmt string, v ...interface{})
  Error(brief, detailsFmt string, v ...interface{}) 
}
```

Then, we can implement two Loggers:

```go
type JUnitLogger struct {}
type ConsoleLogger struct {}
```

And the runner would accept multiple Loggers. Then, it would emit Logging events to each:

```go
log.Warning // in runner calls:
  (*JUnitLogger).Warning
  (*ConsoleLogger).Warning
```

Finally, the main.go file would initialize and set the correct loggers on the Runner itself? This addresses (3) in https://github.com/grpc/test-infra/pull/157#discussion_r637596992, since a JUnitLogger and its data structures would never be created unless the JUnit XML reporter flags are present.

paulosjca commented on inspect/doc.go:
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright 2020 gRPC authors.
```
2021.

codeblooded commented:
It could call both by using a `LoggerList` type:

```go
type LoggerList struct {
  loggers []Logger
}

func (ll *LoggerList) Info(detailsFmt string, v ...interface{}) {
  for _, logger := range ll.loggers {
    logger.Info(detailsFmt, v...)
  }
}

// likewise for Warning and Error
```

Then the runner can accept and use this. It also allows us to support other formats easily in the future?

codeblooded commented:
We actually don't even need `LoggerList` to be a struct. We can alias the slice and add the receivers so it is a Logger itself:

```go
type LoggerList []Logger

func (ll *LoggerList) Info(detailsFmt string, v ...interface{}) {
  for _, logger := range ll {
    logger.Info(detailsFmt, v...)
  }
}
```

Then the Runner can continue to accept only a single `Logger`, but we can pass a []Logger since it is also a Logger. I like this. It seems clean.

codeblooded commented:
I'll give it a try. I think it's pretty clean.

paulosjca commented:
> I'll give it a try. I think it's pretty clean.

Let's discuss before you go too far with it.

codeblooded commented:
It is nowhere near finished, but I like this abstraction a lot. It continues the logging structure from #159. But, it really separates the responsibilities of the report from the runner. The runner now knows nothing about the order of the report. It doesn't have to know what the test suites, test suite or test case objects are. So there is a clear distinction between the purpose of each type. The runner uses a single instance of the `Logger` interface. It only needs to call logging functions on it. Under the hood, a slice of Logger(s) `LoggerList` also implements the `Logger` interface. It proxies the calls to each of the Logger(s) in the slice. Then we only add a `JUnitLogger` to the slice when the JUnit flags are passed. It can be configured and changed without modifying the `Runner`. No matter what, we create a `TextLogger` which logs to stdout. The runtime efficiency is identical, since the `JUnitLogger` uses maps for ordering under the hood.

I still need to address the other feedback. I was just getting some structure together with interfaces that could work.

paulosjca commented:
> I still need to address the other feedback. I was just getting some structure together with interfaces that could work. This structure also makes it easy to add additional output types in the future.

In the meantime, PTAL https://github.com/grpc/test-infra/pull/166.

paulosjca commented:
> > I still need to address the other feedback. I was just getting some structure together with interfaces that could work. This structure also makes it easy to add additional output types in the future.
> 
> In the meantime, PTAL #166.

The idea of the logger LGTM. In the current state of the PR, it looks like the logger is storing data to generate the report (with maps, etc). It might make more sense to store the data in the form of the report itself (as I tried to do in my PR). So you create the report object before creating the logger, the logger fills it in, then you use a method of the report itself to print it out as bytes.

codeblooded commented on tools/runner/runner.go:
```diff
@@ -97,14 +101,13 @@ func (r *Runner) runTest(config *grpcv1.LoadTest, reporter *TestCaseReporter, do
 				continue
 			}
-			reporter.Error("Aborting after %d retries to create test %s: %v", r.retries, name, err)
+			reporter.Error("Error creating the test", "Aborting test %q after %d retries to create test %q failed: %v", name, r.retries, name, err)
```
@paulosjca I tried to clean things up a bit, but when we output the name of the test seemed a little confusing to me. I adopted the feedback and created a rule for myself:

- [x] Parenthesis are no longer used. They are replaced with commas.
- [x] Retries and normal status updates do not include the name of the test.
- [x] Test names are not quoted.
- [x] The test name is only included in an error where all retries have failed. This is consistent with line 107 in the diffbase.

codeblooded commented:
Addressed most feedback items. As discussed, will create a Report object that the JUnitLogger will use through dependency injection in cmd/runner/main.go. Once working, will mark this as Ready for Review.

codeblooded commented:
I liked the Logger implementation, but I found that there was too many places I had to modify the runner and main executable. For this reason, I wound up with a design closer to #159. I'm not a huge fan of many nil-checks in the reporters. So, I'd much rather us use a design pattern that can remove those. For example, it may be nice to modify this to have a plugin architecture (observer pattern) with one plugin being JUnit. This would remove the nil-checks b/c we just wouldn't add the JUnit plugin as an observer. That said, I think this accomplishes the task for now and would like to move on until a plugin system would be necessary (i.e. supporting other report types).

paulosjca commented:
Please re-request review when ready. Thanks.

codeblooded commented:
Showing that ResultStore still works after structural changes: https://source.cloud.google.com/results/invocations/9ca1b2d8-398f-4ea8-a592-d4c22b28bf1e/targets;collapsed=/github%2Fgrpc/tests;group=workers-8core;test=benreed-bfc5a4e9-8404-5132-8ac8-760b500d70f6;row=20.

codeblooded commented on tools/runner/runner.go:
```diff
@@ -41,6 +42,22 @@ type Runner struct {
 	retries        uint
+	startTime      time.Time
+	suites         []*TestSuite
+	mux            sync.Mutex
```
A mutex lock provides a way for the report to be kept up-to-date in real-time without contention, so a DOM traversal is not required at the end. The current version includes a DOM traversal and drops the mutex. But, it's an interesting tradeoff to consider. Generally the mutex only guards integer increments.

codeblooded commented:
@jtattermusch @paulosjca 

I will generate a failing report as requested tomorrow.

As for test case names, do we want them to match the Kubernetes LoadTest names? I ask because the Kubernetes LoadTest names will contain a lot of information, including prefixes, timestamps, etc. I think the only thing that really needs to be displayed as part of the name in the XML report is the scenario name. The rest of the data may be better suited as properties. wdyt?

I will follow-up this PR with ones to add support to capture logs and to produce multiple XML files.

I am marking this as ready for comments.

paulosjca reviewed (approved):
Thanks for doing this. We can fine-tune the report once it is generated continuously.
